File: DesignPatterns/python/Creational/factory.py
"""
Intent: Provides an interface for creating objects in a superclass but
        allows its subclasses to alter the type of object that will be created.
"""

# It allows us to create objects by specifying their common interface at run time without
# imposing their concrete class creation logic.


from __future__ import annotations
from abc import ABC, abstractmethod
from typing import List


class Person:
    def __int__(self, fName, lName):
        self._firstName = fName
        self._lastName = lName
        self.accountDetails = None


class BankAccount(ABC):
    @abstractmethod
    def validateUserIdentity(self):
        pass

    @abstractmethod
    def createAccount(self):
        pass


class BusinessAccount(BankAccount):
    def validateUserIdentity(self):
        print("validating user business identity")

    def createAccount(self):
        print("creating business account")


class CheckinAccount(BankAccount):
    def validateUserIdentity(self):
        print("validating user checkin identity")

    def createAccount(self):
        print("creating checkin account")


def NewAccount(accountType) -> BankAccount:
    account: BankAccount = None

    if accountType == "CheckinAccount":
        account = CheckinAccount()
    elif accountType == "BusinessAccount":
        account = BusinessAccount()

    account.validateUserIdentity()
    account.createAccount()

    return account


if __name__ == "__main__":
    NewAccount("CheckinAccount")

File: DesignPatterns/GO/Creational/factory.go
package main

import "fmt"

func main() {
	account := CheckingAccountFactory("Shiva ")
	account.validateIdentity()
	account.createAccount()

	fmt.Println("")

	account = SavingsAccountFactory("Charan ")
	account.validateIdentity()
	account.createAccount()
}

type BankAccount interface {
	validateIdentity()
	createAccount()
}

type Person struct {
	name string
}

type CheckingAccount struct {
	Person
}

func CheckingAccountFactory(name string) BankAccount {
	return &CheckingAccount{
		Person{
			name: name,
		},
	}
}

func (c *CheckingAccount) validateIdentity() {
	fmt.Printf("validating identity of %s", c.name)
}

func (c *CheckingAccount) createAccount() {
	fmt.Printf("careating  checkin account for %s", c.name)
}

type SavingsAccount struct {
	Person
}

func SavingsAccountFactory(name string) BankAccount {
	return &SavingsAccount{
		Person{
			name: name,
		},
	}
}

func (s *SavingsAccount) validateIdentity() {
	fmt.Printf("validating identity of %s for creating savings account", s.name)
}

func (s *SavingsAccount) createAccount() {
	fmt.Printf("careating  savings account for %s", s.name)
}
